plugins/synaptics-cape/fu-synaptics-cape-device.c

```diff
@@ -0,0 +1,648 @@
+						10000,
+						error);
+}
+
+static gboolean
+fu_synaptics_cape_device_write_firmware_image(FuSynapticsCapeDevice *self,
+					      GBytes *fw,
+					      GError **error)
+{
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	g_return_val_if_fail(FU_IS_SYNAPTICS_CAPE_DEVICE(self), FALSE);
+	g_return_val_if_fail(fw, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	chunks = fu_chunk_array_new_from_bytes(fw,
+					       0x00,
+					       0x00,
+					       FU_SYNAPTICS_CAPE_WORD_IN_BYTES *
+						   FU_SYNAPTICS_CAPE_CMD_WRITE_DATAL_LEN);
+
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks, i);
+		if (!fu_synaptics_cape_device_sendcmd(
+			self,
+			FU_SYNAPTICS_CAPE_CMD_APP_ID('C', 'T', 'R', 'L'),
+			FU_SYNAPTICS_CMD_FW_UPDATE_WRITE,
+			(const guint32 *)fu_chunk_get_data(chk),
+			fu_chunk_get_data_sz(chk) / FU_SYNAPTICS_CAPE_WORD_IN_BYTES,
+			/*fixme*/ 1000 * 200,
+			error)) {
+			g_prefix_error(error, "failed send on chk %u: ", i);
+			return FALSE;
+		}
+		fu_device_set_progress_full(FU_DEVICE(self), i, chunks->len - 1);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_cape_device_write_firmware(FuDevice *device,
+					FuFirmware *firmware,
+					FwupdInstallFlags flags,
+					GError **error)
+{
+	FuSynapticsCapeDevice *self = FU_SYNAPTICS_CAPE_DEVICE(device);
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GBytes) fw_header = NULL;
+
+	g_debug("syna: fu_synaptics_cape_device_write_firmware()\n");
+	fw_header = fu_firmware_get_image_by_id_bytes(firmware, FU_FIRMWARE_ID_HEADER, error);
+	if (fw_header == NULL)
+		return FALSE;
+	if (!fu_synaptics_cape_device_write_firmware_header(self, fw_header, error)) {
+		g_prefix_error(error, "update  header failed: ");
+		return FALSE;
+	}
+
+	/* perform the actual write */
+	fu_device_set_status(device, FWUPD_STATUS_DEVICE_WRITE);
+
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+	if (!fu_synaptics_cape_device_write_firmware_image(self, fw, error)) {
+		g_prefix_error(error, "update image failed: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_cape_device_attach(FuDevice *device, GError **error)
+{
+	FuSynapticsCapeDevice *self = FU_SYNAPTICS_CAPE_DEVICE(device);
+	g_autoptr(GError) error_local = NULL;
+
+	if (fu_device_get_update_state(self) != FWUPD_UPDATE_STATE_SUCCESS)
+		return TRUE;
+
+	if (!fu_synaptics_cape_device_reset(self, error)) {
+		return FALSE;
+	}
+
+	/* wait for re-enumeration */
+	fu_device_set_status(device, FWUPD_STATUS_DEVICE_RESTART);
+	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_cape_device_set_report(FuSynapticsCapeDevice *self,
+				    const guint8 *data,
+				    guint datasz,
+				    GError **error)
+{
+	if (g_getenv("FWUPD_SYNAPTICS_CAPE_HID_REPORT_VERBOSE") != NULL) {
+		fu_common_dump_raw(G_LOG_DOMAIN, "SetReport", data, datasz);
+	}
```

Comment:
no need for the braces when it's a simple one-line contitional

Reply:
okay